Tone/source/ExternalInput.js

```diff
@@ -0,0 +1,235 @@
+define(["Tone/core/Tone", "Tone/source/Source"], function(Tone){
+
+	"use strict";
+
+	/**
+	 *  @class  Tone.ExternalInput is a WebRTC Audio Input. Check 
+	 *          [Media Stream API Support](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream_API)
+	 *          to see which browsers are supported. As of
+	 *          writing this, Chrome, Firefox, and Opera 
+	 *          support Media Stream. Chrome allows enumeration 
+	 *          of the sources, and access to device name over a 
+	 *          secure (HTTPS) connection. See [https://simpl.info](https://simpl.info/getusermedia/sources/index.html) 
+	 *          vs [http://simple.info](https://simpl.info/getusermedia/sources/index.html) 
+	 *          on a Chrome browser for the difference.
+	 *         
+	 *  @constructor
+	 *  @extends {Tone.Source}
+	 *  @param {number} [inputNum=0] If multiple inputs are present, select the input number. Chrome only.
+	 *  @example
+	 *  var motu = new Tone.ExternalInput(3);
+	 *  
+	 *  motu.open(function(){
+	 *  	motu.start(10);
+	 *  });
+	 */
+
+	Tone.ExternalInput = function(){
+
+		var options = this.optionsObject(arguments, ["inputNum"], Tone.ExternalInput.defaults);
+		Tone.Source.call(this, options);
+
+		/**
+		 *  @type {MediaStreamAudioSourceNode}
+		 *  @private
+		 */
+		this._mediaStream = null;
+		
+		/**
+		 *  @type {LocalMediaStream}
+		 *  @private
+		 */
+		this._stream = null;
+		
+		/**
+		 *  @type {Object}
+		 *  @private
+		 */
+		this._constraints = {"audio" : true};
+
+		if(Tone.ExternalInput.sourceList.length === 0){
+			Tone.ExternalInput.getSources(function(){});
+		}
+
+		/**
+		 * The input source position in Tone.ExternalInput.sourceList. 
+		 * Set before ExternalInput.open().
+		 * @type {[type]}
+		 */
+		this.inputNum = options.inputNum;
+
+	};
+
+	Tone.extend(Tone.ExternalInput, Tone.Source);
+
+	/**
+	 * the default parameters
+	 * @type {Object}
+	 */
+	Tone.ExternalInput.defaults = {
+		"inputNum" : 0
+	};
+
+	/**
+	 * wrapper for getUserMedia function
+	 * @private
+	 */
+	Tone.ExternalInput.prototype._getUserMedia = function(callback){
+		this._constraints = {
+			audio : {
+				optional : [{sourceId: Tone.ExternalInput.sourceList[this.inputNum].id}]
+			}
+		};
+		navigator.getUserMedia(this._constraints, function(stream){
+			this._onStream(stream);
+			callback();
+		}.bind(this), function(err){
+			callback(err);
+		});
+	};
+
+	/**
+	 *  Open the media stream 
+	 *  @param  {function =} callback The callback function to 
+	 *                       execute when the stream is open
+	 *  @return {Tone.ExternalInput} this
+	 */
+	Tone.ExternalInput.prototype.open = function(callback){
+		if(Tone.ExternalInput.sourceList.length === 0){
+			Tone.ExternalInput.getSources(function(){
+				this._getUserMedia(callback);
+			}.bind(this));
+		} else {
```

Comment:
could be simplified:

Tone.ExternalInput.getSources(function(){
  this._getUserMedia(callback);
}.bind(this));